# Importing the data and libraries

In [126]:
#importing useful libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime
from sklearn.linear_model import LinearRegression

In [4]:
#importing the csv arquives
precos = pd.read_csv("precos.csv", sep = ";")
pesos = pd.read_csv("pesos.csv", sep = ";")

# Question 2A

In [5]:
#Filling empty values (NA) with the last value of the series before the NA value.
precos.fillna(method="ffill", inplace = True)
pesos.fillna(method="ffill", inplace = True)

In [33]:
# Dropping useless columns and storing in the new dataframe "precos2"
precos2 = precos.drop(precos.columns[14:], axis = 1)

# Dropping useless columns and storing in the new dataframe "precos2"
precos2 = precos2.drop(precos.columns[:1], axis = 1)


#setting the date as the index of the dataframe
precos2.set_index(precos['date'], inplace = True)

# Convert historical stock prices to daily percent change
price_change = precos2.pct_change(1)

# Deletes row one containing the NaN and stores it in the DataFrame "df"
df = price_change.drop(price_change.index[0])

In [76]:
# this function prints the beta for two variables using linear regression

def regression_model(x, y):
    x1 = np.array(x).reshape((-1,1))
    y1 = np.array(y)
    model = LinearRegression().fit(x1, y1)
    return model.coef_

In [79]:
# iteration
start_time = datetime.now()
for i in df.columns[:11]:
    tt = regression_model(df["Benchmark Mercado"], df[i])
    print("The Beta of {0} and the benchmark is {1}".format(i,tt))
    
end_time = datetime.now()
print('Duration: {}'.format(end_time - start_time))

The Beta of AAAA3 and the benchmark is [1.348806]
The Beta of BBBB3 and the benchmark is [1.38961938]
The Beta of CCCC4 and the benchmark is [0.93501354]
The Beta of DDDD3 and the benchmark is [0.75744315]
The Beta of EEEE3 and the benchmark is [0.68156847]
The Beta of FFFF3 and the benchmark is [0.87394279]
The Beta of GGGG3 and the benchmark is [0.89583935]
The Beta of ZZZZ3 and the benchmark is [0.87428725]
The Beta of YYYY3 and the benchmark is [0.94811451]
The Beta of WWWW3 and the benchmark is [0.97420797]
The Beta of HHHH3 and the benchmark is [0.14753134]
Duration: 0:00:00.007977


# Question 2B

The linear regression is a way to obserse how two variables stands together in a line in a certain period of time, how they variate with each other. A beta of 1.348806 means that the asset AAAA3 is 34.8806% more volatile than the benchmark. Another way of looking at this is that with 1% variance in the benchmark, the AAAA3 asset will vary 1.348806% on average.

# Question 2C

In [115]:
def beta(x,y):
    covariance = sum((x - x.mean())*((y- y.mean())))/(len(x)-1)
    variance_x = sum((x - x.mean())**2)/(len(x)-1)
    
    return(covariance/variance_x)

In [117]:
start_time = datetime.now()

for i in df.columns[:11]:
    tt = beta(df["Benchmark Mercado"], df[i])
    print("The Beta of {0} and the benchmark is {1}".format(i,tt))
    
end_time = datetime.now()
print('Duration: {}'.format(end_time - start_time))

The Beta of AAAA3 and the benchmark is 1.348806003518077
The Beta of BBBB3 and the benchmark is 1.3896193797436778
The Beta of CCCC4 and the benchmark is 0.935013542314543
The Beta of DDDD3 and the benchmark is 0.7574431486426462
The Beta of EEEE3 and the benchmark is 0.6815684697414911
The Beta of FFFF3 and the benchmark is 0.8739427863290586
The Beta of GGGG3 and the benchmark is 0.8958393493095255
The Beta of ZZZZ3 and the benchmark is 0.8742872534995356
The Beta of YYYY3 and the benchmark is 0.9481145065995564
The Beta of WWWW3 and the benchmark is 0.9742079727274086
The Beta of HHHH3 and the benchmark is 0.1475313351205279
Duration: 0:00:00.010941


# Question 2D

The technique used in item A is more efficient than the one used in item C. In other words, the computational cost of the technique used in item A is bigger, measured by the response time calculated for each item. A more efficient algorithm is always better; less energy is used, and it is more easy to scale, combining it with other algorithms and processing large amounts of data. 

# Question 2E